--- SDL2-2.0.12/src/core/android/SDL_android.h	2020-03-11 09:36:18.000000000 +0800
+++ termux-sdl/app/src/main/jni/SDL2/src/core/android/SDL_android.h	2020-07-29 22:27:19.975245674 +0800
@@ -137,6 +137,24 @@
 void Android_ActivityMutex_Unlock(void);
 void Android_ActivityMutex_Lock_Running(void);
 
+// 设置亮度
+void SDL_AndroidSetBrightness(int);
+// 获取亮度
+int SDL_AndroidGetBrightness(void);
+
+// 设置音量
+void SDL_AndroidSetVolume(int);
+// 获取音量
+int SDL_AndroidGetVolume(void);
+
+// 打印log到java层
+void SDL_AndroidLogPrint(int, const char*, const char*, ...)
+#if defined(__GNUC__)
+    __attribute__((__format__(printf, 3, 4)))
+#endif
+; // SDL_AndroidLogPrint
+
+
 /* Ends C function definitions when using C++ */
 #ifdef __cplusplus
 /* *INDENT-OFF* */
